### PR TITLE
AO3-4473 Skin wizard improvements

### PR DESCRIPTION
**Pull request by sarken**
https://otwarchive.atlassian.net/browse/AO3-4473

Just gonna copy from the issue:

1. The fields are all in one jumbled mess. Let's break them out into sensible chunks, e.g. a section for basic information, another for fonts, and another for colors.

2. Provide more information about the fields on the form instead of making users click the help text... but not by putting in in the label. For instance, instead of the field label saying, "Foreground color (name or hex code)," the label should be "Foreground color" and there should be a footnote-style message beneath the field telling users to format it as a name or hex code. If we use aria-describedby, it'll be read for screenreaders, too.

3. For clarity and accuracy, some labels should be changed. "Foreground color" label should be "Text color", "Font family" should be "Font", "Margin width" should be "Work margin width", and "Percent of base font size" should be "Percent of browser font size". The text on the skin's show page should match.

4. Entering a "Font" should do a better job of changing the font. Right now, for example, it fails to change the font on form buttons, e.g. Subscribe, Favorite Tag, or the font on the header's links. (Note: This can continue to change only the main font. Anything in Georgia, e.g. headings and the site name, should remain in Georgia. In the future, we can give users an ~accent font option for those.)

5. Entering a "Header color" should change the red parts of the header. Right now, it changes the white part. /o\

6. Entering an "Accent color" should change most of the grey parts, e.g. forms and the dropdown menus in the header. Right now it changes the dashboard, meta, and behind the icon in the header.

7. Entering a "Work margin width" should actually adjust the work margins in a fairly predictable way. Right now it really doesn't because of the max-width on the work area.

8. Entering a "Percent of browser font size" should increase or decrease all font sizes on the Archive by that percentage. Right now... idk. (This name is still a little misleading -- as you may know, most, but not all, of the Archive doesn't use your browser's font size, but rather a size that's ~87.5% of the default. Other name suggestions are appreciated.)

9. If a skin was created with the wizard, the edit link should drop you directly in the wizard.

10. If you create a skin with the wizard, you should also be able to add custom CSS to the same skin to override or tweak it, e.g. if you make the "Header color" a light pink and don't change anything else, you can choose "Edit," switch to "Write Custom CSS", and enter `#header .heading a { color: purple }` to make the site name more readable on the white background. Right now, if you try to add custom CSS to a wizard skin, it completely overrides the wizard skin.

*Before this is deployed to staging,* make and apply a skin with the current incarnation of the wizard. This pull request includes an after task that will let you keep your existing wizard skin(s) as they now look. Why you'd _want_ to, I don't know, but...

### CODE REVIEW

File: app/models/skin.rb [hunk 1]
```diff
@@ -127,6 +128,17 @@ def clean_css
   scope :unapproved_skins, where(:public => true, :official => false, :rejected => false)
   scope :rejected_skins, where(:public => true, :official => false, :rejected => true)
   scope :site_skins, where(:type => nil)
+  scope :wizard_site_skins, where("type IS NULL AND (
```
Comment (sarken) [inline app/models/skin.rb]: This is only used for the after task at the moment, but it seems potentially useful include in the model. If others disagree, I'll move it.

File: app/models/skin.rb [hunk 2]
```diff
@@ -184,6 +196,10 @@ def byline
     end
   end
 
+  def wizard_settings?
```
Comment (sarken) [inline app/models/skin.rb]: Just so we can check if a given skin has wizard settings.

File: app/models/skin.rb [hunk 3]
```diff
@@ -353,10 +357,13 @@ def get_style_block(roles_to_include)
       if roles_to_include.include?(get_role)
         if self.filename.present?
           block += get_ie_comment(stylesheet_link(self.filename, get_media))
-        elsif self.css.present?
```
Comment (sarken) [inline app/models/skin.rb]: We were kind of assuming that a skin could have CSS _or_ have wizard settings, but not both, and only including the custom CSS. Now what we do is add the wizard styles followed by the custom CSS you might've added to tweak the wizard settings.

File: app/views/skins/_skin_navigation.html.erb [hunk 4]
```diff
@@ -20,7 +20,13 @@
     <% end %>
         
     <% if skin.editable? %>
-      <li><%= span_if_current ts('Edit'), edit_skin_url(skin) %></li>
+      <li>
```
Comment (sarken) [inline app/views/skins/_skin_navigation.html.erb]: If the skin has wizard settings, just go straight to the wizard editor.

File: app/views/skins/_skin_style_block.html.erb [hunk 5]
```diff
@@ -1,28 +1,4 @@
-<!-- expects local "skin" and is cached one level up -->
+<% # expects local "skin" and is cached one level up %>
 <style type="text/css">
```
Comment (sarken) [inline app/views/skins/_skin_style_block.html.erb]: This partial is only used for work skins. The reset of the stuff had no reason to be there.

File: public/stylesheets/site/2.0/11-group-listbox.css [hunk 6]
```diff
@@ -22,7 +22,7 @@ http://otwcode.github.com/docs/front_end_coding/patterns/listbox
 
 .listbox > .heading, .listbox .heading a:visited {
   margin: 0;
-  color: #222;
+  color: #2a2a2a;
```
Comment (sarken) [inline public/stylesheets/site/2.0/11-group-listbox.css]: We don't actually use `#222` anywhere else, which became evident while collecting elements with like colors for the wizard skin.

**Comment by sarken** [comment]
Have to update this because #2359 caused some conflicts.

File: app/models/skin.rb (hunk 3 above)
Comment (cesy) [inline app/models/skin.rb]: I like this.

File: app/models/skin.rb [hunk 7]
```diff
@@ -353,10 +357,13 @@ def get_style_block(roles_to_include)
...
-          block += get_ie_comment('<style type="text/css" media="' + get_media + '">' + self.css + '</style>')
-        elsif (wizard_block = get_wizard_settings).present?
-          block += '<style type="text/css" media="' + get_media + '">' + wizard_block + '</style>'
+        else
+          if (wizard_block = get_wizard_settings).present?
```
Comment (cesy) [inline app/models/skin.rb]: That's a lovely dream, hound.

File: app/views/skins/_skin_navigation.html.erb (hunk 4 above)
Comment (cesy) [inline app/views/skins/_skin_navigation.html.erb]: I really like these usability improvements.

File: app/views/skins/_skin_style_block.html.erb [hunk 8]
```diff
@@ -1,28 +1,4 @@
-<!-- expects local "skin" and is cached one level up -->
+<% # expects local "skin" and is cached one level up %>
```
Comment (cesy) [inline app/views/skins/_skin_style_block.html.erb]: Hooray for code tidying.

File: app/views/skins/_wizard_form.html.erb [hunk 9]
```diff
@@ -1,59 +1,110 @@
...
+      </p>
+    </dd>
+
+    <dt>
+      <%= f.label :accent_color, ts('Accent color') %>
+    </dt>
+    <dd>
+      <%= f.text_field :accent_color, "aria-describedby" => "accent-color-field-notes" %>
+      <p class="footnote" id="accent-color-field-notes">
+        <%= ts("Name or hex code. Default: <code>#fff</code>".html_safe) %>
```
Comment (cesy) [inline app/views/skins/_wizard_form.html.erb]: Putting the defaults for all of these is great. Is it also worth explaining what we mean by accent colour?

File: lib/skin_wizard.rb [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+          margin: auto #{percentage}%;
+          max-width: 100%;
+        }
+      "
+    else
+      ""
+    end
+  end
+end
+
```
Comment (cesy) [inline lib/skin_wizard.rb]: This is probably worth tidying.

File: public/help/skins-wizard-vertical-gap.html [hunk 11]
```diff
@@ -1,5 +1,3 @@
-﻿<p>The default is: 1.125</p>
-<p>Put any number in here and it will be counted as a percentage of the 
-vertical gaps between paragraphs. 
-A larger number will give you bigger spacing between paragraphs.</p>
-<p>1.5 is fairly big spacing, 2 is huge, 0.5 is small.</p>
\ No newline at end of file
+﻿<p>The default is: <code>1.1286em</code></p>
+<p>Put any number in here and it will be applied as a multiple of the work's  font size. A larger number will give you bigger spacing between paragraphs.</p>
```
Comment (cesy) [inline public/help/skins-wizard-vertical-gap.html]: Double space in "work's font"

File: public/stylesheets/site/2.0/02-elements.css [hunk 12]
```diff
@@ -67,7 +67,6 @@ h2 {
   line-height: 1;
   margin: 0.429em 0;
   display: inline;
-    text-shadow: -1px -1px 2px #fff;
```
Comment (cesy) [inline public/stylesheets/site/2.0/02-elements.css]: Was this change meant to be in here? It seems unrelated to the rest and possibly needs different testing.

File: app/views/skins/_wizard_form.html.erb (hunk 9 above)
Comment (sarken) [inline app/views/skins/_wizard_form.html.erb]: I'm game for detailing what the colors are for, although I think I'll put it in as help text, since the explanations might get a little too long. (Also, note to self: the default here is wrong. This is a grey, despite what the old help text said.)

**Comment by cesy** [comment]
Two typos and a couple of questions, but the rest of this looks sensible. This is one of the types of changes that you can spend a long time looking at the code and still miss stuff, and sadly it's also the type where no matter how much testing LO does, we'll still miss stuff that a user will notice. But I've had a good go through it and I think this is the best we'll get, unless @bingeling has loads of time to double-check it.

File: public/stylesheets/site/2.0/02-elements.css (hunk 12 above)
Comment (sarken) [inline public/stylesheets/site/2.0/02-elements.css]: Yeah, it's actually a shadow that's invisible against the white background (we neglected to take it out after we got rid of the blue/lilac/grey background), but that leads to ugliness or readability issues once you start changing the background and font colors.

**Comment by sarken** [comment]
Updated! I can't get Hound's complaint about the missing newline to go away for love or money, but the little red "no newline" character is gone, so ¯\_(ツ)_/¯
